### PR TITLE
Avatar eye look at fixes for HMD and desktop

### PR DESCRIPTION
**Pull request by hyperlogic**
Several fixes for HMD and desktop eye tracking.

* You should see eye 'saccades' when viewing other avatars looking at you, while you are in desktop mode.
* Your eye look ats will continue to work even if you are in mirror mode.
* Your eyes should move with your head if there are no look at targets, and you are wearing an HMD.

### CODE REVIEW

**Comment by hyperlogic** [comment]
https://app.asana.com/0/45663370065758/94885064383503

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -864,36 +864,26 @@ void MyAvatar::updateLookAtTargetAvatar() {
...
+                    glm::vec3 rightEyeHeadLocal = glm::vec3(rightEye[3]);
+                    auto humanSystem = qApp->getViewFrustum();
+                    glm::vec3 humanLeftEye = humanSystem->getPosition() + (humanSystem->getOrientation() * leftEyeHeadLocal);
+                    glm::vec3 humanRightEye = humanSystem->getPosition() + (humanSystem->getOrientation() * rightEyeHeadLocal);
+
+                    // Scale by proportional differences between avatar and human.
+                    float humanEyeSeparationInModelSpace = glm::length(humanLeftEye - humanRightEye);
+                    float avatarEyeSeparation = glm::length(avatarLeftEye - avatarRightEye);
+                    gazeOffset = gazeOffset * humanEyeSeparationInModelSpace / avatarEyeSeparation;
+                }
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: When in desktop mode, the eye offsets were zero.  This was causing the eye cycling between the left, right eyes and mouth to not function.  So, now we only scale the gazeOffset if you are in HMD mode.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -2946,9 +2946,9 @@ void Application::updateMyAvatarLookAtPosition() {
         } else {
             //  I am not looking at anyone else, so just look forward
             if (isHMD) {
-                glm::mat4 headPose = _avatarUpdate->getHeadPose() ;
+                glm::mat4 headPose = _avatarUpdate->getHeadPose();
                 glm::quat headRotation = glm::quat_cast(headPose);
-                lookAtSpot = _myCamera.getPosition() +
+                lookAtSpot = myAvatar->getPosition() +
                     myAvatar->getOrientation() * (headRotation * glm::vec3(0.0f, 0.0f, -TREE_SCALE));
```
Comment (hyperlogic) [inline interface/src/Application.cpp]: When looking "forward" the lookAtSpot should be from the avatar's position, not the camera.

File: interface/src/avatar/MyAvatar.cpp [hunk 3]
```diff
@@ -849,7 +849,7 @@ void MyAvatar::updateLookAtTargetAvatar() {
         avatar->setIsLookAtTarget(false);
         if (!avatar->isMyAvatar() && avatar->isInitialized() &&
             (distanceTo < GREATEST_LOOKING_AT_DISTANCE * getUniformScale())) {
-            float angleTo = glm::angle(lookForward, glm::normalize(avatar->getHead()->getEyePosition() - cameraPosition));
+            float angleTo = glm::angle(lookForward, glm::normalize(avatar->getHead()->getEyePosition() - getHead()->getEyePosition()));
             if (angleTo < (smallestAngleTo * (isCurrentTarget ? KEEP_LOOKING_AT_CURRENT_ANGLE_FACTOR : 1.0f))) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: For determining facing we should use the vector between the two heads, not between the avatar's head and the camera.

File: libraries/animation/src/Rig.cpp [hunk 4]
```diff
@@ -1067,14 +1067,21 @@ void Rig::updateEyeJoint(int index, const glm::vec3& modelTranslation, const glm
...
 
         // limit rotation
         const float MAX_ANGLE = 30.0f * RADIANS_PER_DEGREE;
-        q = glm::angleAxis(glm::clamp(glm::angle(q), -MAX_ANGLE, MAX_ANGLE), glm::axis(q));
+        deltaQuat = glm::angleAxis(glm::clamp(glm::angle(deltaQuat), -MAX_ANGLE, MAX_ANGLE), glm::axis(deltaQuat));
 
         // directly set absolutePose rotation
-        _internalPoseSet._absolutePoses[index].rot = q;
+        _internalPoseSet._absolutePoses[index].rot = deltaQuat * headQuat;
     }
```
Comment (hyperlogic) [inline libraries/animation/src/Rig.cpp]: We now clamp the delta quaternion from the head to the eye, rather than the eye quaternion itself.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 460582239a804a2a4dc069e30e7ebf5a50217537 available at: https://gist.github.com/b161b816a0a6ba548ba6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.dmg

File: interface/src/avatar/AvatarUpdate.cpp [hunk 5]
```diff
@@ -31,7 +31,14 @@ void AvatarUpdate::synchronousProcess() {
...
     auto frameCount = qApp->getFrameCount();
-    _headPose = qApp->getActiveDisplayPlugin()->getHeadPose(frameCount);
+
+    QSharedPointer<AvatarManager> manager = DependencyManager::get<AvatarManager>();
+    MyAvatar* myAvatar = manager->getMyAvatar();
+    assert(myAvatar);
+
+    // transform the head pose from the displayPlugin into avatar coordinates.
+    glm::mat4 invAvatarMat = glm::inverse(createMatFromQuatAndPos(myAvatar->getOrientation(), myAvatar->getPosition()));
+    _headPose = invAvatarMat * (myAvatar->getSensorToWorldMatrix() * qApp->getActiveDisplayPlugin()->getHeadPose(frameCount));
```
Comment (hyperlogic) [inline interface/src/avatar/AvatarUpdate.cpp]: This looks convoluted, but it's taking the HMD from sensor space into avatar space.  It's like a scene graph with three nodes:
<pre>
    (world)
      /\
     /  \
    /    \
 (room) (avatar)
   |
   |
   |
 (hmd)
</pre>

To get the hmd into avatar space.  X = avatar<sup>-1</sup> * room * hmd

File: libraries/animation/src/Rig.cpp [hunk 6]
```diff
@@ -1067,14 +1067,21 @@ void Rig::updateEyeJoint(int index, const glm::vec3& modelTranslation, const glm
...
         glm::vec3 zAxis = glm::normalize(_internalPoseSet._absolutePoses[index].trans - transformPoint(worldToRig, lookAtSpot));
-        glm::quat q = rotationBetween(IDENTITY_FRONT, zAxis);
+
+        glm::quat desiredQuat = rotationBetween(IDENTITY_FRONT, zAxis);
+        glm::quat headQuat;
+        int headIndex = _animSkeleton->nameToJointIndex("Head");
+        if (headIndex >= 0) {
+            headQuat = _internalPoseSet._absolutePoses[headIndex].rot;
+        }
+        glm::quat deltaQuat = desiredQuat * glm::inverse(headQuat);
```
Comment (howard-stearns) [inline libraries/animation/src/Rig.cpp]: Initialize headQuat and/or log if no "Head" joint? (To not get silent garbage.)

File: interface/src/avatar/MyAvatar.cpp (hunk 1 above)
Comment (howard-stearns) [inline interface/src/avatar/MyAvatar.cpp]: How do these relate to the existing getIPD and getIPDScale?  Would they help simplify? (I don't know. Just asking.)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3cde9721743da428f4dfc25aaeb34f1dde3d9e5e available at: https://gist.github.com/b161b816a0a6ba548ba6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: babf48ac5756131734cf6c2bcc0cae4a2a660f00 available at: https://gist.github.com/b161b816a0a6ba548ba6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 506e46faeeaa3f7d16dc9f6eaffe4c5c116b3a85 available at: https://gist.github.com/b161b816a0a6ba548ba6<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7227.dmg

**Comment by cozza13** [comment]
I went through the tests on this and it works

**Pull request merged**